File: src/lib/langfuse-import/import-service.ts
/**
 * Langfuse import service — orchestrates full and incremental data imports.
 * Server-only module.
 *
 * Usage:
 *   const service = new LangfuseImportService();
 *   const result = await service.runImport();
 *
 * The service auto-detects whether to run a full or incremental import
 * based on whether prior successful imports exist in the database.
 */
import "server-only";

import { LangfuseRepository } from "./repository";
import { LangfuseClient } from "./langfuse-client";
import type {
  ImportRecord,
  DailyMetric,
  LangfuseRequest,
  LangfuseApiDailyData,
  LangfuseApiObservation,
} from "./types";

/** Number of days to look back for a full import. */
const FULL_IMPORT_LOOKBACK_DAYS = 90;

export class LangfuseImportService {
  private repo: LangfuseRepository;
  private client: LangfuseClient;

  constructor(repo?: LangfuseRepository, client?: LangfuseClient) {
    this.repo = repo ?? new LangfuseRepository();
    this.client = client ?? new LangfuseClient();
  }

  /**
   * Runs a Langfuse import. Auto-detects full vs incremental mode.
   *
   * - Full import: triggered when no prior successful import exists.
   *   Fetches the last 90 days of daily metrics and all observations.
   *
   * - Incremental import: triggered when a prior successful import exists.
   *   Fetches only data newer than the last import's to_timestamp.
   *
   * The import is idempotent — duplicate data is upserted, not duplicated.
   * On failure, the import run is marked as 'failed' with the error message.
   *
   * @returns The completed ImportRecord (status will be 'success' or 'failed').
   */
  async runImport(): Promise<ImportRecord> {
    const lastImport = this.repo.getLastSuccessfulImport();
    const isIncremental = lastImport !== null;

    const now = new Date();
    const toTimestamp = now.toISOString();

    let fromTimestamp: string | null = null;
    let fromDate: string;

    if (isIncremental && lastImport) {
      // Incremental: start from last successful import's to_timestamp
      fromTimestamp = lastImport.to_timestamp;
      fromDate = lastImport.to_timestamp.split("T")[0];
    } else {
      // Full: look back N days
      const lookback = new Date(now);
      lookback.setDate(lookback.getDate() - FULL_IMPORT_LOOKBACK_DAYS);
      fromDate = lookback.toISOString().split("T")[0];
    }

    const toDate = now.toISOString().split("T")[0];
    const mode = isIncremental ? "incremental" : "full";

    // Create the import run record
    const importRun = this.repo.createImportRun(
      mode,
      fromTimestamp,
      toTimestamp,
    );

    try {
      // 1. Fetch and upsert daily metrics
      const rawMetrics = await this.client.fetchDailyMetrics(fromDate, toDate);
      const dailyMetrics = this.transformDailyMetrics(rawMetrics);
      this.repo.upsertDailyMetrics(dailyMetrics);

      // 2. Fetch and upsert observations/requests
      const rawObservations = await this.client.fetchAllObservations(
        fromTimestamp ?? undefined,
      );
      const requests = this.transformObservations(rawObservations);
      this.repo.upsertRequests(requests);

      // 3. Mark import as successful
      this.repo.completeImportRun(importRun.id, "success");

      return {
        ...importRun,
        finished_at: new Date().toISOString(),
        status: "success",
      };
    } catch (err) {
      const errorMessage =
        err instanceof Error ? err.message : String(err);
      this.repo.completeImportRun(importRun.id, "failed", errorMessage);

      return {
        ...importRun,
        finished_at: new Date().toISOString(),
        status: "failed",
        error_message: errorMessage,
      };
    }
  }

  /**
   * Transforms raw Langfuse v1 Metrics API data into flat DailyMetric rows.
   * Each row in the API response is already one (date, model) combination.
   */
  private transformDailyMetrics(
    raw: LangfuseApiDailyData[],
  ): DailyMetric[] {
    return raw.map((row) => ({
      date: row.time_dimension.split("T")[0],
      model: row.providedModelName ?? "unknown",
      input_tokens: Number(row.sum_inputTokens) || 0,
      output_tokens: Number(row.sum_outputTokens) || 0,
      total_tokens: Number(row.sum_totalTokens) || 0,
      request_count: Number(row.count_count) || 0,
      total_cost: row.sum_totalCost ?? 0,
    }));
  }

  /**
   * Transforms raw Langfuse observation API data into LangfuseRequest rows.
   */
  private transformObservations(
    raw: LangfuseApiObservation[],
  ): LangfuseRequest[] {
    return raw.map((obs) => {
      const inputTokens = obs.usage?.input ?? 0;
      const outputTokens = obs.usage?.output ?? 0;
      const totalTokens = obs.usage?.total ?? 0;

      let latencyMs: number | null = null;
      if (obs.startTime && obs.endTime) {
        latencyMs =
          new Date(obs.endTime).getTime() -
          new Date(obs.startTime).getTime();
      }

      return {
        id: obs.id,
        trace_id: obs.traceId ?? null,
        name: obs.name ?? null,
        model: obs.model ?? null,
        started_at: obs.startTime ?? null,
        finished_at: obs.endTime ?? null,
        input_tokens: inputTokens,
        output_tokens: outputTokens,
        total_tokens: totalTokens,
        cost: obs.calculatedTotalCost ?? null,
        latency_ms: latencyMs,
      };
    });
  }
}

File: services/api/app/observability/application/import_service.py
import logging
from datetime import datetime, timedelta, timezone

from app.observability.application.ports import LangfuseClientPort, LangfuseRepositoryPort
from app.observability.domain.models import DailyMetric, ImportRecord, LangfuseRequest

logger = logging.getLogger(__name__)

FULL_IMPORT_LOOKBACK_DAYS = 90


class ImportService:
    def __init__(self, repo: LangfuseRepositoryPort, client: LangfuseClientPort) -> None:
        self._repo = repo
        self._client = client

    async def run_import(self) -> dict:
        last_import = await self._repo.get_last_successful_import()
        from_timestamp, from_date, to_date, to_timestamp, mode = self._resolve_import_range(
            last_import
        )

        import_run = await self._repo.create_import_run(mode, from_timestamp, to_timestamp)

        try:
            raw_metrics = await self._client.fetch_daily_metrics(from_date, to_date)
            daily_metrics = self._transform_daily_metrics(raw_metrics)
            await self._repo.upsert_daily_metrics(daily_metrics)

            raw_observations = await self._client.fetch_all_observations(from_timestamp)
            requests = self._transform_observations(raw_observations)
            await self._repo.upsert_requests(requests)

            await self._repo.complete_import_run(import_run.id, "success")

            return {
                **_import_record_to_dict(import_run),
                "finished_at": datetime.now(timezone.utc).isoformat(),
                "status": "success",
            }
        except Exception as err:
            logger.exception("Langfuse import failed")
            error_message = str(err)
            await self._repo.complete_import_run(import_run.id, "failed", error_message)
            raise

    @staticmethod
    def _resolve_import_range(
        last_import: ImportRecord | None,
    ) -> tuple[str | None, str, str, str, str]:
        """Return (from_timestamp, from_date, to_date, to_timestamp, mode)."""
        now = datetime.now(timezone.utc)
        to_timestamp = now.isoformat()
        to_date = now.isoformat().split("T")[0]

        if last_import is not None:
            from_timestamp: str | None = last_import.to_timestamp
            from_date = last_import.to_timestamp.split("T")[0]
            mode = "incremental"
        else:
            from_timestamp = None
            lookback = now - timedelta(days=FULL_IMPORT_LOOKBACK_DAYS)
            from_date = lookback.isoformat().split("T")[0]
            mode = "full"

        return from_timestamp, from_date, to_date, to_timestamp, mode

    @staticmethod
    def _transform_daily_metrics(raw: list[dict]) -> list[DailyMetric]:
        return [
            DailyMetric(
                date=row.get("time_dimension", "").split("T")[0],
                model=row.get("providedModelName") or "unknown",
                input_tokens=int(row.get("sum_inputTokens") or 0),
                output_tokens=int(row.get("sum_outputTokens") or 0),
                total_tokens=int(row.get("sum_totalTokens") or 0),
                request_count=int(row.get("count_count") or 0),
                total_cost=row.get("sum_totalCost") or 0,
            )
            for row in raw
        ]

    @staticmethod
    def _transform_observations(raw: list[dict]) -> list[LangfuseRequest]:
        results: list[LangfuseRequest] = []
        for obs in raw:
            latency_ms: int | None = None
            if obs.get("latency") is not None:
                latency_ms = round(obs["latency"] * 1000)
            elif obs.get("startTime") and obs.get("endTime"):
                start = datetime.fromisoformat(obs["startTime"].replace("Z", "+00:00"))
                end = datetime.fromisoformat(obs["endTime"].replace("Z", "+00:00"))
                latency_ms = int((end - start).total_seconds() * 1000)

            results.append(
                LangfuseRequest(
                    id=obs["id"],
                    trace_id=obs.get("traceId"),
                    name=obs.get("name"),
                    model=obs.get("model"),
                    started_at=obs.get("startTime"),
                    finished_at=obs.get("endTime"),
                    input_tokens=obs.get("inputUsage") or 0,
                    output_tokens=obs.get("outputUsage") or 0,
                    total_tokens=obs.get("totalUsage") or 0,
                    cost=obs.get("totalCost"),
                    latency_ms=latency_ms,
                )
            )
        return results


def _import_record_to_dict(record: ImportRecord) -> dict:
    return {
        "id": record.id,
        "started_at": record.started_at,
        "finished_at": record.finished_at,
        "mode": record.mode,
        "from_timestamp": record.from_timestamp,
        "to_timestamp": record.to_timestamp,
        "status": record.status,
        "error_message": record.error_message,
    }
